### ISSUE TITLE
Interactive chat for explaining terms and question

### ISSUE THREAD

**Issue by kaljarv** [OpenVAA/voting-advice-application#627]
# Primary user story

As a voter unfamiliar with a VAA statement,
I want to receive answers to questions I have about the statement,
so that I can make a more informed choice when answering.

# Considerations

## Structured or free chat?

It is probably easier for a user to start the chat if there are preprovided questions or highlighted terms they can tap on instead of them having to formulate a freeform query. Such a structured approach provides the additional benefit that the questions and answers can precomputed, approved and edited by the administrator.

If the chat is structured, it may be useful to precompute follow-up questions up to some level as well.

However, voters might have a questions about the statements that are not included in the precomputed set, and in these cases a free chat would be useful.

# Subtasks

1. Initial planning
  - Which technologies to use?
  - Which of the tech is shared between the other ideas?
  - Which data sources are needed?
  - Is the AI model interacted with in real-time or do we use canned results?
  - Make a sketch of the ideal interaction? E.g. a script of the chat or a rough draft of the visualisation
2. Prototyping
  - Make a UI sketch in [Figma](https://www.figma.com/design/3LDNZWOxuhhVaVPAOApfOK/Voting-Advice-Application?node-id=4169-15907&t=o7Z4IrwYPkeGbBeH-1)
  - Test the suggested workflows using real data and real models
3. Development
  - Structured 'canned' chat
  - Generate:
    - Introduction to question
    - Explanations for all terms in the statement and the introduction
    - What is the current situation
    - Ability to add more generatable items
  - Localizable
  - Voter UI
    - Common with #628
4. Optional admin UI development
  - Edit introductions and 'current situation' summaries
  - Edit, delete and add term explanations
    - Automatically link these to the terms in the introductions without explicit a-elements?
  - Localizable
  - Perhaps just an export/import of tabular or JSON data

# Update 2025-01-28

## Strapi endpoints

- Define what new Strapi endpoints are needed (read/write)
- Write functions will have to run on server only to be able to use an API key stored in `env` variables (or in the future use a new authenticated admin user type)

## Settings

An example of the settings to add for the new functions.

For now it's okay if these are only added to `packages/app-shared` and not implemented in Strapi’s `App Settings`.

```ts
// In app-shared: dynamicSettings.ts
interface DynamicSettings {
  // ... existing stuff
  questions: {  
    // ... existing stuff
    interactiveInfo?: {
      /**
       * Default `false`.
       */
      enabled?: boolean;
      // Possible other settings
    }
  }
}
```

## Admin functions

See [Figma](https://www.figma.com/proto/3LDNZWOxuhhVaVPAOApfOK/Voting-Advice-Application?node-id=4550-16628&t=gXzXO8KXLNhBvJ0N-1&scaling=scale-down&content-scaling=fixed&page-id=0%3A1&starting-point-node-id=4550%3A16628&show-proto-sidebar=1) for a preliminary design.

**NB** User login can be implemented later.

### Interactive Question Info

- `generateQuestionInfo(questionIds?: Array<Id>): Promise<{ type: 'success' | 'failure' }>`
  - If `questionIds` is specified, perform for only those. Otherwise all.
  - Generate for all app languages.

- Edit question info
  - Perhaps for now just use a ready-made JSON editor like https://github.com/josdejong/svelte-jsoneditor
  - Enable export/import as CSV or other tabular format

# Update 2025-02-19: Term explanation popups

The term explanations are complicated by the fact that the terms may appear in an inflected form in the question, e.g.:

- Question: "Valtiontalouden pitäisi olla tasapainossa."
- Term: "Valtiontalous"

We don't want to edit the original question text by, e.g. adding some tags there, so we would also need to save for each term the actual string that triggers the term popup – and add the popup in runtime to the text displayed. This should happen in the [`QuestionHeading`](https://github.com/OpenVAA/voting-advice-application/blob/feat-misc-improvements/frontend/src/lib/dynamic-components/questionHeading/QuestionHeading.svelte) component which is part of the update due to be merged this week.

This may demand a data structure like this for the terms. Furthermore, it might be nice to harmonize the structure for all of the info content to follow the same model:

```ts
// The defs below are multilingual i.e. the format in which they would be saved in Strapi. 
// In the frontend all localizations would be translated into a single locale by `StrapiDataProvider`.

type TermDefinition = ContentItem & {
  /** The strings that trigger the popup. There may in theory be many different forms in the same question. */
  triggers: {[locale: string]: Array<string>};
}

interface ContentItem = {
  /** The title of the item. */
  title: LocalizedString;
  /** The text or HTML content of the item. */
  content: LocalizedString;
}
```

**Comment by edthu** [comment]
[Notion page](https://www.notion.so/Interactive-chat-for-explaining-terms-and-question-13040660a20a8080ae6feec3ed98341f) for the solutions.

**Comment by kaljarv** [comment]
Check "Update 2025-02-19: Term explanation popups" in the main description.

**Comment by kaljarv** [comment]
> Hey [@kaljarv](https://github.com/kaljarv), did I understand correctly that you'd like the term explanation ui to be based off of the feat-combined-features-for-vaa-2025-02 branch? If so, any eta on when will it be merged?

Yes, that'd be preferable because the actual triggers for the terms popups could be baked in to the `QuestionHeading` component. We should have the review of that branch ready today, so it could be merged tomorrow. Sorry for the delay…

Ofc, if you keep that in mind, you can start work already.

> Also, should the term popups be modals under the hood?

I think that'd be a bit of overkill for them. Maybe you can start from: https://daisyui.com/components/tooltip/

And ig the component could be called `Tooltip` accordingly.

**Comment by edthu** [comment]
Hi @kaljarv has there been an authentication method decided for the admin ui? I am creating the endpoints and wondering what kind of policy we would like for the route.

**Comment by kaljarv** [comment]
> Hi [@kaljarv](https://github.com/kaljarv) has there been an authentication method decided for the admin ui? I am creating the endpoints and wondering what kind of policy we would like for the route.

I think it would be easiest to create a new `Admin` user role in the Users Permissions extension and create a global policy `user-is-admin` which checks whether the call is authenticated and has that role. 

The permissions details are currently quite messily handled in `/backend/vaa-strapi/src/extensions/users-permissions/strapi-server.ts`. Note that you'll need to also edit the default permission generated so that `Admin` is granted all the public API permissions as well, i.e., everything that starts with `api::`.

The policy would in turn go to `/backend/vaa-strapi/src/policies`.
